### PR TITLE
gh-125196: Use PyUnicodeWriter for repr(list)

### PR DESCRIPTION
**Pull request by vstinner**
Replace the private _PyUnicodeWriter with the public PyUnicodeWriter.

* Issue: gh-125196

### CODE REVIEW

File: Objects/listobject.c [hunk 1]
```diff
@@ -522,49 +522,45 @@ list_dealloc(PyObject *self)
...
     }
 
-    _PyUnicodeWriter_Init(&writer);
-    writer.overallocate = 1;
     /* "[" + "1" + ", 2" * (len - 1) + "]" */
-    writer.min_length = 1 + 1 + (2 + 1) * (Py_SIZE(v) - 1) + 1;
+    Py_ssize_t prealloc = 1 + 1 + (2 + 1) * (Py_SIZE(v) - 1) + 1;
+    PyUnicodeWriter *writer = PyUnicodeWriter_Create(prealloc);
+    if (writer == NULL) {
+        return NULL;
```
Comment (kumaraditya303) [inline Objects/listobject.c]: You need to call `Py_ReprLeave` before returning `NULL`

Comment (vstinner) [reply Objects/listobject.c]: Nice catch, fixed!

**Comment by vstinner** [comment]
Benchmark:

```py
import pyperf
runner = pyperf.Runner()

runner.bench_func('repr([])', repr, [])
runner.bench_func('repr([1,2,3])', repr, [1, 2, 3])
list3 = ['abcdef']*10
runner.bench_func('repr(list3)', repr, list3)
list4 = ['abcdef']*50
runner.bench_func('repr(list4)', repr, list4)
```

Result, Python built with `gcc -O3`:

```
+----------------+---------+-----------------------+
| Benchmark      | ref     | change                |
+================+=========+=======================+
| repr([])       | 104 ns  | 98.1 ns: 1.06x faster |
+----------------+---------+-----------------------+
| repr([1,2,3])  | 534 ns  | 558 ns: 1.04x slower  |
+----------------+---------+-----------------------+
| repr(list3)    | 1.53 us | 1.59 us: 1.04x slower |
+----------------+---------+-----------------------+
| repr(list4)    | 5.36 us | 5.53 us: 1.03x slower |
+----------------+---------+-----------------------+
| Geometric mean | (ref)   | 1.01x slower          |
+----------------+---------+-----------------------+
```

In the worst case, it's 24 ns slower, 534 ns => 558 ns: 1.04x slower.

@serhiy-storchaka @pitrou: Do you think that it's an acceptable slowdown, under 10% slower on a microbenchmark?

File: Objects/listobject.c [hunk 2]
```diff
@@ -522,49 +522,47 @@ list_dealloc(PyObject *self)
...
+    }
 
     Py_ReprLeave((PyObject *)v);
-    return _PyUnicodeWriter_Finish(&writer);
+    return PyUnicodeWriter_Finish(writer);
 
 error:
-    _PyUnicodeWriter_Dealloc(&writer);
+    if (writer != NULL) {
+        PyUnicodeWriter_Discard(writer);
```
Comment (serhiy-storchaka) [inline Objects/listobject.c:567]: Should not `PyUnicodeWriter_Discard(NULL)` be no-op?

File: Objects/listobject.c [hunk 3]
```diff
@@ -522,49 +522,47 @@ list_dealloc(PyObject *self)
...
         goto error;
+    }
 
     /* Do repr() on each element.  Note that this may mutate the list,
        so must refetch the list size on each iteration. */
-    for (i = 0; i < Py_SIZE(v); ++i) {
+    for (Py_ssize_t i = 0; i < Py_SIZE(v); ++i) {
         if (i > 0) {
-            if (_PyUnicodeWriter_WriteASCIIString(&writer, ", ", 2) < 0)
+            if (PyUnicodeWriter_WriteUTF8(writer, ", ", 2) < 0) {
```
Comment (serhiy-storchaka) [inline Objects/listobject.c]: What is faster, `PyUnicodeWriter_WriteUTF8(writer, ", ", 2)` or two `PyUnicodeWriter_WriteChar()`?

This is an idle question. Even if the latter is marginally faster, it may still be more preferable to use the former. But if the difference is significant, this may be a question for optimization.

**Comment by pitrou** [comment]
> @serhiy-storchaka @pitrou: Do you think that it's an acceptable slowdown, under 10% slower on a microbenchmark?

Definitely!

**Comment by vstinner** [comment]
> Definitely!

Combined with a minor change, PR gh-125214, it's even *faster* :-)

File: Objects/listobject.c (hunk 3 above)
Comment (vstinner) [reply Objects/listobject.c]: > What is faster, PyUnicodeWriter_WriteUTF8(writer, ", ", 2) or two PyUnicodeWriter_WriteChar()?

Two PyUnicodeWriter_WriteChar() calls is faster:

```
+----------------+-------------+-----------------------+
| Benchmark      | change_utf8 | change_2char          |
+================+=============+=======================+
| repr([])       | 50.2 ns     | 54.0 ns: 1.07x slower |
+----------------+-------------+-----------------------+
| repr(list3)    | 825 ns      | 802 ns: 1.03x faster  |
+----------------+-------------+-----------------------+
| repr(list4)    | 2.89 us     | 2.66 us: 1.09x faster |
+----------------+-------------+-----------------------+
| Geometric mean | (ref)       | 1.01x faster          |
+----------------+-------------+-----------------------+
```

File: Objects/listobject.c (hunk 2 above)
Comment (vstinner) [reply Objects/listobject.c:567]: Ok, I modified `PyUnicodeWriter_Discard(writer)` to do nothing if writer is NULL.

**Comment by serhiy-storchaka** [comment]
> Combined with a minor change, PR [gh-125214](https://github.com/python/cpython/pull/125214), it's even _faster_ :-)

Well, than integers are not suitable for this microbenchmark, you should use something else.

File: Objects/unicodeobject.c [hunk 4]
```diff
@@ -13428,6 +13428,9 @@ PyUnicodeWriter_Create(Py_ssize_t length)
 
 void PyUnicodeWriter_Discard(PyUnicodeWriter *writer)
 {
+    if (writer == NULL) {
```
Comment (serhiy-storchaka) [inline Objects/unicodeobject.c]: This should be documented. Please do this in a separate PR.

Also, I think it is worth to try to optimize `PyUnicodeWriter_WriteUTF8()` for short ASCII strings.

Comment (vstinner) [reply Objects/unicodeobject.c]: > This should be documented. Please do this in a separate PR.

Oh ok. I reverted the change.

**Comment by vstinner** [comment]
Updated benchmark results with CPU isolation:

```
vstinner@mona$ python3 -m pyperf compare_to --table ref.json change.json
+----------------+---------+-----------------------+
| Benchmark      | ref     | change                |
+================+=========+=======================+
| repr([])       | 57.2 ns | 52.2 ns: 1.09x faster |
+----------------+---------+-----------------------+
| repr([1,2,3])  | 283 ns  | 293 ns: 1.03x slower  |
+----------------+---------+-----------------------+
| repr(list3)    | 827 ns  | 807 ns: 1.02x faster  |
+----------------+---------+-----------------------+
| repr(list4)    | 2.80 us | 2.74 us: 1.02x faster |
+----------------+---------+-----------------------+
| Geometric mean | (ref)   | 1.03x faster          |
+----------------+---------+-----------------------+
```

Now it's only slower for `repr([1,2,3])`, but this row will be made way faster using gh-125214.

> Well, than integers are not suitable for this microbenchmark, you should use something else.

Only `repr([1,2,3])` row uses benchmarks, list3 and list4 use strings.

**Comment by vstinner** [comment]
I merged, thanks for reviews. There is apparently room for performance improvement. Let's use this change as a starting point.
